Add basic configuration system

sdilts opened the pull request:
This PR adds a basic configuration utility package that covers all of the essential elements of a configuration system as described in #838. Before merging this commit, look at https://github.com/szos/defcustom/tree/functionize, which @szos is writting to solve the same problem. Comparing the two, it looks like szos's version has many more features and nice-to-haves, while the code here just meets the minimal requirements and is much simpler.

## Features

This PR adds the macros `define-config-parameter` and `define-config-var` for declaring variables that are meant to be modified by the system users. These macros register the variables and declare them using `defparameter` and `defvar` respectively. A validation function is given to this macro that is used for checking that the variable's value conforms to specific requirements. `define-config-enum` is also present, which works the same way as `define-config-parameter` except the the validation function is generated automatically.

To set configuration variables, you can use the `set-config` and `set-config-atomic` macros. These macros follow the same form as `setf` and verify that the values the variables are being set to are correct using the provided validation function. If the values aren't valid, an `invalid-datum-error` condition is raised. The difference between these two functions are that if an error condition is raised while `set-configuration-atomic` is being executed, all of the variables in form are reset to their values from before the form was executed.

You can retrieve information about a specific configuration variable by using the `get-configuration-info` function, and get a list of all of the known configuration variables with `all-config-info`. You can print information about individual variables or variables matching a pattern to a stream using `describe-config-info` and `describe-all-config-info`, respectively.

A `with-atomic-update` macro is also provided, which resets a given set of variables to their values before the form was executed if a condition was raised while the form was executing.

Since the code uses `defparameter` and `defvar` under the hood, it is completely backwards compatible with existing code.

sdilts commented on config-system.lisp:
```diff
@@ -0,0 +1,157 @@
+  ;; if we are given a lambda, we need to evaluate it in order to call it
+  (%generate-config-var-code 'defvar name default validator documentation))
+
+(defmacro define-config-enum (name default values documentation)
```
There are no tests for this macro.

sdilts commented on config-system.lisp:
```diff
@@ -0,0 +1,157 @@
+  `(define-config-parameter ,name ,default (lambda (x) (member x ,values :test #'equal))
+                            ,documentation))
+
+(defun list-all-configurations ()
```
I'm not a big fan of this function name. Maybe `get-all-config-info`?

szos replied:
What has been recommended to me (i think by puercopop) was to avoid using `get-` in getters. I think a name like `all-config-info` would be better. As mentioned in another response, this function could also benefit from a database optional or key variable, if we go that route.

sdilts commented on config-system.lisp:
```diff
@@ -0,0 +1,157 @@
+(defun get-configuration-info (setting-name)
+  "Find the info for the config variable stored in the symbol SETTING-NAME."
+  (declare (type symbol setting-name))
+  (gethash setting-name *config-vars*))
```
The functions have the hash table hard-coded in, which will probably lead to problems in the future.

sdilts replied:
What you are talking about is something that [`destructring-bind`](http://clhs.lisp.se/Body/m_destru.htm) provides. In this case though, I think an optional parameter with the default of `*config-vars*` would be fine.

sdilts commented on config-system.lisp:
```diff
@@ -0,0 +1,157 @@
+                                     :validator ,func :doc ,documentation))
+             (error 'invalid-datum-error :place (quote ,name) :value ,default-value))))))
+
+(defmacro define-config-parameter (name default validator documentation)
```
Using keyword parameters here might be nice to allow for future expansion.

szos replied:
I agree with using keywords - I think it would be useful for things like changing the default database(s) or providing a different test function for `define-config-enum`, for example.

PuercoPop replied:
Yeah, in particular I won't have separate macros for parameter/var, instead use keyword argument `:reinitialize` to use either defvar or defparameter.

I would also move documentation to a keyword argument and use the ['unsplice'](https://github.com/ruricolist/serapeum/blob/master/REFERENCE.md#unsplice-form) idiom ej. `,@(when documentation (list documentation))`. Or do we want to enforce a docstring for every configuration option?

> `define-config-enum`

I think that the valid values are better left as a keyword argument not a new kind of config parameter. ej. `(defconfig theme :type '(member 'dark 'light))`

szos replied:
I lean towards having documentation be optional, as by requiring it the parameters/vars that wouldnt have them would most likely just put an empty string, which feels wrong. This is purely personal opinion, not necessarily informed by logic.

sdilts replied:
> I wouldn't have separate macros for parameter/var, instead use keyword argument :reinitialize to use either defvar or defparameter.

Good idea!

> Or do we want to enforce a docstring for every configuration option?

This was my intention, but it looks like I'm outvoted. Considering these are settings, there should be some type of blurb to go along with it. I can see the merits of not requiring it, and like the interface of using a keyword better anyways. I'll take it out.

> I think that the valid values are better left as a keyword argument not a new kind of config parameter

That does feel more natural. So instead of specifying a validation function, you would specify a [type specifier](http://www.lispworks.com/documentation/HyperSpec/Body/04_bc.htm) that would be used to make a new type? A validation method is required though, so I'm hesitant to put that in a keyword argument, although it is pretty easy to signal an error when one isn't included.

szos replied:
In [defconfig](https://github.com/szos/defconfig) I've been using just a quoted list of items (or unquoted if its a function to generate such a list) which gets spliced into a call to member for the `:valid-values` key argument. I like the idea of using type specifiers, but id like to have the ability to just pass in a list of acceptable values alongside it. I see this being useful for things like `*input-window-gravity*` and the like which are just a list of keywords.

side note - I would love any and all feedback on defconfig, as its my first real system with a real purpose, and there are definitely design flaws.

sdilts replied:
The[`member`](http://www.lispworks.com/documentation/HyperSpec/Body/t_member.htm)type specifier takes care of that, as long as the values can be compared with `eql`. You also have the `satisfies` specifier, which can take any predicate function.

szos replied:
Ah! this thread has been quite educating for me! love it :smile:

PuercoPop replied:
> This was my intention, but it looks like I'm outvoted.

I was just asking questions :grimacing: . I may have brought up the pattern out of reflex/custom. But upon thinking about it, forcing a docstring is a valid choice to make sure we don't slack with the documentation. I say we keep it as obligatory. 

> So instead of specifying a validation function, you would specify a type specifier that would be used to make a new type?

StumpWM already has its own deftype macro, define-stumpwm-type. One way to go about it would be to extend it to support validators. But that would be on a separate PR.

sdilts reviewed (commented):
Here are some potential improvements / problems with the provided implementation.

szos commented on config-system.lisp:
```diff
@@ -0,0 +1,157 @@
+  (declare (type symbol setting-name))
+  (gethash setting-name *config-vars*))
+
+(defmacro %set-configuration (setting-name value)
```
I think it would be a good idea to store/dispatch based not only on the symbol but on the symbols package as well. This macro and the underlying hash table don't appear to currently support this. I solved this by making the hash table use `#'equalp` and using `(package . symbol)` for the hash key. What do you think of this? alternatively one could use the full name of the symbol as the key. This work is already partially done within `%full-symbol-string`.

sdilts replied:
Symbols from different packages are distinct items, so there is no need to do this: they will never compare to be equal. For example, while in the `cl-user` package, the form `(eql `sb-ext::%boundp '%boundp)` returns `nil`. So as long as a package isn't using a symbol from another package, symbols with the same name from different packages will be distinct keys in the hash table.

szos replied:
Ah! thanks for explaining this! looks like I have solved a nonexistent problem :rofl:

szos reviewed (commented):
Ive added some feedback regarding features and how things are stored. Hope its useful :smile:

sdilts commented:
I've added some functions to print information on the config variables to a stream, added some filtering options to `all-config-info`, and normalized all of the the names to use `config` rather than have some use `configure`. With those changes, I'm happy with the PR and think that the  code is ready to be merged. That being said, we should decide if we want to be able to specify different config tables instead of the default before moving forward.

szos commented:
I think for this system, since its only (currently) being used by stump, the ability to specify a database isnt absolutely neccessary.

PuercoPop commented:
One thing I'm not clear is, what is the purpose of all the introspection machinery at this stage? If at a later date the intention is to provide a GUI for the settings the introspection can be add in the PR that does that. As long as we are using the define-config-* macro that can be introduced w/o changing the call-sites.

sdilts commented:
> what is the purpose of all the introspection machinery at this stage?

I was thinking that a repl-level interface for searching and getting info for settings would be nice, and help fulfill the "user friendly" and "self documenting" requirements that were laid out in the proposal. I think having some type of interactive interface, even if it through a repl in Slime would be much better than what we currently have. Like you said however, it's not a requirement to get the basics working.

PuercoPop commented:
> Like you said however, it's not a requirement to get the basics working.

From what we have discussed above there are a couple of design decisions that merit further discussion (and hopefully input from more people). In the interest of not loosing steam and to help keep the discussion focused I would prefer to have this merged in smaller increments. What do you about the following plan to move forward:

1. Have on PR where we only replace the calls to `defvar` and `defparameter` with the new macro w/o any change in semantics. This would be straightforward to review and merge.
2. A second PR where we introduce an API for validation to the new macro (and if it makes sense to tie it with `define-stumpwm-type`.
3. Finally a PR to introduce the introspection capabilities.


Btw I used `defconfig` because I'm a lazy typer :grimacing: (and favor that naming variant), but the initial `define-*` naming is more aligned with the current codebase. Feel free to keep the previous name. If we end up going with the shorter name I think `defcustom` would be a better choice. As StumpWM users are likely to be Emacs users and emacs uses defcustom.

Happy new year y'all :partying_face: :tada:

dmb2 commented:
Have we decided on this vs @szos example? I'm happy to back this as long as my concerns from the original proposal are addressed.  Specifically, I think a hashtable in memory of the original values is OK, but a database on disk seems like overkill (unless I'm overlooking something obvious).

szos commented:
I think that, at least for the moment, we should use @sdilts's implementation or a variation of it. The [system I'm working on](github.com/szos/defconfig) is still being tweaked, and is a bit more heavyweight than strictly necessary for StumpWM. I'm also wanting to try using it in a more complicated way than the tests suite ive written, in order to see what design decisions make the most sense - thusly the API may be subject to change (though I feel fairly settled on most of it). 

I'm a fan of @PuercoPop's proposal to start small and work our way up by initially only replacing the defvar/parameter calls with the new macro(s), then adding validation, and finally introspective capabilities. To this end I think using @sdilts's implementation will be easier/more malleable.

dmb2 commented:
fix the conflicts and I'll merge this.

szos commented:
Its been a little under a month with no updates here, Id like to put forward the idea that this could be done as a contrib module. I've started working on such a thing [here](https://github.com/szos/stumpwm-settings), and while it would mean writing settings names twice, it might have less impact than integrating with stumpwm proper. thoughts?

dmb2 commented:
lets do this as a module
